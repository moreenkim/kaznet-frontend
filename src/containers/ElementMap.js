// This component takes a list of strings
// and returns them wrapped in the provided HTML tag
import React, { Component } from "react";
import { DropdownToggle, DropdownMenu, Dropdown } from "reactstrap";
import PropTypes from "prop-types";

export default class ElementMap extends Component {
  static propTypes = {
    filterFields: PropTypes.object
  };

  render() {
    const items = this.props.items;
    // use React.createElement to create the html tag
    // in a dynamic way
    const listItems = items.map((item, index) =>
      React.createElement(
        this.props.HTMLTag,
        { key: index, className: this.props.className },
        [item]
      )
    );

    if (this.props.filterFields != null) {
      /// Check for filterFields prop
      /// If present create a DropDown wrapped inside specified
      /// HTMLTag prop.
      const sortItems = Object.entries(this.props.filterFields).map(
        ([name, dropDownItems], index) => {
          const uniqueKey = 2 * listItems.length;
          return React.createElement(
            this.props.HTMLTag,
            { key: uniqueKey + index, className: this.props.className },
            [
              <Dropdown
                key={uniqueKey * 2}
                isOpen={this.props.isOpen}
                toggle={e => this.props.handleChange(e)}
              >
                <DropdownToggle caret tag="span" className="nav-link">
                  {name}
                </DropdownToggle>
                <DropdownMenu>{dropDownItems}</DropdownMenu>
              </Dropdown>
            ]
          );
        }
      );

      /// Modify listItems to Include Filter Fields in Specified Position in
      /// this.props.filterItemPositions
      for (var i = 0; i < this.props.filterItemPositions.length; i++) {
        listItems.splice(this.props.filterItemPositions[i], 0, sortItems[i]);
      }
    }

    return listItems;
  }
}
