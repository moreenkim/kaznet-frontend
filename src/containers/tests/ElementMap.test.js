// Test Element Map
import React from "react";
import { shallow, mount } from "enzyme";
import ElementMap from "../ElementMap";
import toJson from "enzyme-to-json";

describe("containers/ElementMap", () => {
  it("renders without crashing", () => {
    const headerItems = ["Name"];
    shallow(<ElementMap items={headerItems} HTMLTag="th" />);
  });

  it("renders tags correctly", () => {
    const headerItems = ["Name"];
    const wrapper = mount(<ElementMap items={headerItems} HTMLTag="div" />);

    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.unmount();
  });
});
