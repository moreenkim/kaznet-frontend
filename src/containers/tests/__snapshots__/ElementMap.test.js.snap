// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/ElementMap renders filter field 1`] = `
<ElementMap
  HTMLTag="div"
  filterFilterFields={
    Object {
      "Status": Array [],
    }
  }
  filterItemPositions={
    Array [
      0,
    ]
  }
  items={
    Array [
      "Status",
    ]
  }
>
  <div
    key="0"
  >
    Status
  </div>
</ElementMap>
`;

exports[`containers/ElementMap renders tags correctly 1`] = `
<ElementMap
  HTMLTag="div"
  items={
    Array [
      "Name",
    ]
  }
>
  <div
    key="0"
  >
    Name
  </div>
</ElementMap>
`;
